cloud-controller-manager/do/metrics.go

```diff
@@ -0,0 +1,71 @@
+/*
+Copyright 2020 DigitalOcean
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package do
+
+import "github.com/prometheus/client_golang/prometheus"
+
+type metrics struct {
+	host               string
+	apiRequestDuration *prometheus.HistogramVec
+	runLoopDuration    *prometheus.HistogramVec
+	reconcileDuration  *prometheus.HistogramVec
+}
+
+// width of 4.5 represents 90/20:
+// (timeout duration in seconds) /
+// (the number of desired buckets minus 1 to account for the excluded bucket)
+const width float64 = 4.5
+
+// create metrics
+var (
+	apiRequestDuration = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: "firewall",
+			Name:      "http_request_duration_seconds",
+			Help:      "This is a histogram for tracking the duration of firewall API requests.",
+			Buckets:   prometheus.LinearBuckets(1, width, 21),
+		},
+		[]string{"method", "code"},
+	)
+	runLoopDuration = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: "firewall",
+			Name:      "run_loop_duration_seconds",
+			Help:      "This is a histogram for tracking the duration of the run loop and whether it succeeds or not.",
+			Buckets:   prometheus.LinearBuckets(1, width, 21),
+		},
+		[]string{"success"},
+	)
+	reconcileDuration = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: "firewall",
+			Name:      "reconcile_duration_seconds",
+			Help:      "The duration of time, in seconds, that it takes for the firewall reconcile to run and ensure that the firewall is reconciled.",
+			Buckets:   prometheus.LinearBuckets(1, width, 21),
```

Comment:
I'm not sure what we expect the range to be for `reconcile_duration_seconds` but given that `run_loop_duration_seconds` bucket definitions were copied from `http_request_duration_seconds` I would feel better if you dug in to understand what the expected range should be for that one as well. If you can then confirm that this bucket definition will be fine I'll be :+1: here.

Reply:
yeah this should be different, i'll update. maybe with some comments too

Reply:
Run and Reconcile will have the same range. Because of the fwReconcileFrequency value which is `5 * time.Minute`. So, both `run_loop_duration_seconds` and `reconcile_duration_seconds` should have a width of `300/20` as pointed out in your previous comment. I've updated the code to address this.

Also, I agree. `http_request_duration_seconds` should remain `90 * time.Second`.

Thank you!